vault/seal/transit/transit.go

```diff
@@ -0,0 +1,256 @@
+			return nil, err
+		}
+		if config["token"] != "" {
+			client.SetToken(config["token"])
+		}
+		if namespace != "" {
+			client.SetNamespace(namespace)
+		}
+		if client.Token() == "" {
+			return nil, errors.New("missing token")
+		}
+		s.client = client
+
+		// Send a value to test the seal and to set the current key id
+		if _, err := s.Encrypt(context.Background(), []byte("a")); err != nil {
+			return nil, err
+		}
+
+		if !disableRenewal {
+			// Renew the token immediately to get a secret to pass to renewer
+			secret, err := client.Auth().Token().RenewTokenAsSelf(s.client.Token(), 0)
+			// If we don't get an error renewing, set up a renewer.  The token may not be renewable or not have
+			// permission to renew-self.
+			if err == nil {
+				renewer, err := s.client.NewRenewer(&api.RenewerInput{
+					Secret: secret,
+				})
+				if err != nil {
+					return nil, err
+				}
+				s.renewer = renewer
+
+				go func() {
+					for {
+						select {
+						case err := <-renewer.DoneCh():
+							s.logger.Info("shutting down token renewal")
+							if err != nil {
+								s.logger.Error("error renewing token", "error", err)
+							}
+							return
+						case <-renewer.RenewCh():
+							s.logger.Trace("successfully renewed token")
+						}
+					}
+				}()
+				go s.renewer.Renew()
+			} else {
+				s.logger.Info("unable to renew token, disabling renewal", "err", err)
+			}
+		}
+	}
+
+	sealInfo := make(map[string]string)
+	sealInfo["address"] = s.client.Address()
+	sealInfo["mount_path"] = s.mountPath
+	sealInfo["key_name"] = s.keyName
+	if namespace != "" {
+		sealInfo["namespace"] = namespace
+	}
+
+	return sealInfo, nil
+}
+
+// Init is called during core.Initialize
+func (s *Seal) Init(_ context.Context) error {
+	return nil
+}
+
+// Finalize is called during shutdown
+func (s *Seal) Finalize(_ context.Context) error {
+	if s.renewer != nil {
+		s.renewer.Stop()
+	}
+
+	return nil
+}
+
+// SealType returns the seal type for this particular seal implementation.
+func (s *Seal) SealType() string {
+	return seal.Transit
+}
+
+// KeyID returns the last known key id.
+func (s *Seal) KeyID() string {
+	return s.currentKeyID.Load().(string)
+}
+
+// Encrypt is used to encrypt using Vaults Transit engine
+func (s *Seal) Encrypt(_ context.Context, plaintext []byte) (*physical.EncryptedBlobInfo, error) {
+	encPlaintext := base64.StdEncoding.EncodeToString(plaintext)
+	path := path.Join(s.mountPath, "encrypt", s.keyName)
+	secret, err := s.client.Logical().Write(path, map[string]interface{}{
+		"plaintext": encPlaintext,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	ciphertext := secret.Data["ciphertext"].(string)
+	splitKey := strings.Split(ciphertext, ":")
+	if len(splitKey) != 3 {
+		return nil, errors.New("invalid ciphertext returned")
+	}
+	keyID := splitKey[1]
+	s.currentKeyID.Store(keyID)
+
+	ret := &physical.EncryptedBlobInfo{
+		Ciphertext: []byte(ciphertext),
+		KeyInfo: &physical.SealKeyInfo{
+			KeyID: keyID,
```

Comment:
Should we be storing the keyName as well as the key term here?

Reply:
This very much follows the mold or the other cloud auto unseals, the key that you are using is fixed while the version can change.  The only unseal that doesn't follow this mold is the enterprise HSM integration, but versions are usually not a construct that HSMs support.